test/extensions/filters/udp/udp_proxy/udp_proxy_integration_test.cc

```diff
@@ -0,0 +1,182 @@
+
+  static std::string configToUse() {
+    return ConfigHelper::BASE_UDP_LISTENER_CONFIG + R"EOF(
+    listener_filters:
+      name: envoy.filters.udp_listener.udp_proxy
+      typed_config:
+        '@type': type.googleapis.com/envoy.config.filter.udp.udp_proxy.v2alpha.UdpProxyConfig
+        cluster: cluster_0
+      )EOF";
+  }
+
+  /**
+   * Initializer for an individual test.
+   */
+  void SetUp() override {
+    udp_fake_upstream_ = true;
+    BaseIntegrationTest::initialize();
+  }
+
+  /**
+   *  Destructor for an individual test.
+   */
+  void TearDown() override {
+    test_server_.reset();
+    fake_upstreams_.clear();
+  }
+
+  void requestResponseWithListenerAddress(const Network::Address::Instance& listener_address) {
+    // Send datagram to be proxied.
+    UdpSyncClient client(timeSystem(), version_);
+    client.write("hello", listener_address);
+
+    // Wait for the upstream datagram.
+    Network::UdpRecvData request_datagram;
+    ASSERT_TRUE(fake_upstreams_[0]->waitForUdpDatagram(request_datagram));
+    EXPECT_EQ("hello", request_datagram.buffer_->toString());
+
+    // Respond from the upstream.
+    fake_upstreams_[0]->sendUdpDatagram("world", *request_datagram.addresses_.peer_);
+    Network::UdpRecvData response_datagram;
+    client.recv(response_datagram);
+    EXPECT_EQ("world", response_datagram.buffer_->toString());
+    EXPECT_EQ(listener_address.asString(), response_datagram.addresses_.peer_->asString());
+  }
+};
+
+INSTANTIATE_TEST_SUITE_P(IpVersions, UdpProxyIntegrationTest,
+                         testing::ValuesIn(TestEnvironment::getIpVersionsForTest()),
+                         TestUtility::ipTestParamsToString);
+
+TEST_P(UdpProxyIntegrationTest, HelloWorldOnLoopback) {
+  const uint32_t port = lookupPort("listener_0");
+  auto listener_address = Network::Utility::resolveUrl(
+      fmt::format("tcp://{}:{}", Network::Test::getLoopbackAddressUrlString(version_), port));
+  requestResponseWithListenerAddress(*listener_address);
+}
+
+TEST_P(UdpProxyIntegrationTest, HelloWorldOnNonLocalAddress) {
+  const uint32_t port = lookupPort("listener_0");
+  Network::Address::InstanceConstSharedPtr listener_address;
+  if (version_ == Network::Address::IpVersion::v4) {
+    // Kernel regards any 127.x.x.x as local address.
+    listener_address.reset(new Network::Address::Ipv4Instance(
+#ifndef __APPLE__
+        "127.0.0.3",
+#else
+        "127.0.0.1",
+#endif
+        port));
+  } else {
+    // IPv6 doesn't allow any non-local source address for sendmsg. And the only
+    // local address guaranteed in tests in loopback. Unfortunately, even if it's not
+    // specified, kernel will pick this address as source address. So this test
+    // only checks if IoSocketHandle::sendmsg() sets up CMSG_DATA correctly,
+    // i.e. cmsg_len is big enough when that code path is executed.
+    listener_address.reset(new Network::Address::Ipv6Instance("::1", port));
```

Comment:
Does this paragraph indicate that it only tests listening on non-localhost ipV4 address on Linux? If so, can we explicitly say so?

Reply:
I copied this from the other test that I think you wrote. :) I will look more at this.